vpr/src/route/connection_box_lookahead_map.cpp

```diff
@@ -785,91 +756,188 @@ float ConnectionBoxMapLookahead::get_expected_cost(
         VTR_ASSERT(seg_index < num_segments);
 
-        bounding_box_for_segment[seg_index] |= bounding_box_for_node(node);
+        bounding_box_for_segment[seg_index].expand_bounding_box(bounding_box_for_node(device_ctx.connection_boxes, i));
     }
 
-    // select an inode near the center of the bounding box for each segment type
-    inodes_for_segment->clear();
-    inodes_for_segment->resize(num_segments);
-    for (auto& grid : *inodes_for_segment) {
-        for (int y = 0; y < SAMPLE_GRID_SIZE; y++) {
-            for (int x = 0; x < SAMPLE_GRID_SIZE; x++) {
-                grid.point[y][x].samples = std::vector<ssize_t>();
-            }
-        }
+    // initialize counts
+    for (int seg = 0; seg < num_segments; seg++) {
+        const auto& box = bounding_box_for_segment[seg];
+        segment_counts[seg] = vtr::Matrix<int>({size_t(box.width()), size_t(box.height())}, 0);
     }
 
+    // count sample points
     for (size_t i = 0; i < rr_nodes.size(); i++) {
         auto& node = rr_nodes[i];
-        vtr::Rect<int> node_bounds = bounding_box_for_node(node);
         if (node.type() != CHANX && node.type() != CHANY) continue;
-        if (node.capacity() == 0 || device_ctx.connection_boxes.find_canonical_loc(i) == nullptr) continue;
+        if (node.capacity() == 0 || node.num_edges() == 0) continue;
+        const std::pair<size_t, size_t>* loc = device_ctx.connection_boxes.find_canonical_loc(i);
+        if (loc == nullptr) continue;
 
         int seg_index = device_ctx.rr_indexed_data[node.cost_index()].seg_index;
+        segment_counts[seg_index][loc->first][loc->second] += 1;
 
         VTR_ASSERT(seg_index != OPEN);
         VTR_ASSERT(seg_index < num_segments);
+    }
 
-        auto& grid = (*inodes_for_segment)[seg_index];
-        for (int sy = 0; sy < SAMPLE_GRID_SIZE; sy++) {
-            for (int sx = 0; sx < SAMPLE_GRID_SIZE; sx++) {
-                auto& point = grid.point[sy][sx];
-                if (point.samples.empty()) {
-                    point.samples.push_back(i);
-                    point.location = vtr::Point<int>(node.xlow(), node.ylow());
-                    goto next_rr_node;
+    // select sample points
+    for (int i = 0; i < num_segments; i++) {
+        const auto& counts = segment_counts[i];
+        const auto& bounding_box = bounding_box_for_segment[i];
+        if (bounding_box.empty()) continue;
+        for (int y = 0; y < SAMPLE_GRID_SIZE; y++) {
+            for (int x = 0; x < SAMPLE_GRID_SIZE; x++) {
+                vtr::Rect<int> window = sample_window(bounding_box, x, y, SAMPLE_GRID_SIZE);
+                auto histogram = count_histogram(window, segment_counts[i]);
+                SampleRegion region = {
+                    /* .segment_type = */ i,
+                    /* .grid_location = */ vtr::Point<int>(x, y),
+                    /* .points = */ choose_points(counts, window, quantile(histogram, 0.5), quantile(histogram, 0.7)),
```

Comment:
Move the `0.5` and `0.7` to named comments (e.g. constexpr double kXxxxYyyy), with a brief comment about how you picked them.

Reply:
Done.